synapse/storage/database.py

```diff
@@ -461,8 +461,23 @@ def new_transaction(
         exception_callbacks: List[_CallbackListEntry],
         func: "Callable[..., R]",
         *args: Any,
+        db_retry: bool = True,
         **kwargs: Any
     ) -> R:
+        """Start a new database transaction with the given connection.
+
+        Args:
+            conn
+            desc
+            after_callbacks
+            exception_callbacks
+            func
+            *args
+            db_retry: Whether to retry the transaction by calling `func` again.
+                This should be disabled if connection is in autocommit mode.
```

Comment:
```suggestion
            db_retry: Whether to retry the transaction after an OperationalError by calling `func` again.
                This should be disabled if connection is in autocommit mode.
```

Reply:
the problem with disabling this is that it's the thing that deals with postgres connections dropping (which can happen for annoying TCP reasons). I'm not sure it's safe to do: why is it required?

Reply:
Unlike functions that are run in transactions its not necessarily true its safe to just re-run transaction functions, since they might have done half the work already. I guess we can mandate that such functions should be safe to re-run? (All the usages are for example)

Reply:
I've done that.